src/gt4py/next/program_processors/runners/dace/gtir_to_sdfg.py

```diff
@@ -509,26 +581,42 @@ def add_nested_sdfg(
 
         input_memlets = {}
         unused_data = set()
-        for nsdfg_dataname, nsdfg_datadesc in inner_ctx.sdfg.arrays.items():
-            if nsdfg_datadesc.transient:
+        for data, datadesc in inner_ctx.sdfg.arrays.items():
+            if datadesc.transient:
                 pass  # nothing to do here
-            elif nsdfg_dataname in data_args:
-                arg_node = data_args[nsdfg_dataname]
-                source_data = arg_node.dc_node.data
-                input_memlets[nsdfg_dataname] = outer_ctx.sdfg.make_array_memlet(source_data)
-            elif nsdfg_dataname in outer_ctx.sdfg.arrays:
-                # include fields from the parent scope
-                source_data = nsdfg_dataname
-                # ensure that connectivity tables are non-transient arrays in parent SDFG
-                if source_data in connectivity_arrays:
-                    outer_ctx.sdfg.arrays[source_data].transient = False
-                input_memlets[nsdfg_dataname] = outer_ctx.sdfg.make_array_memlet(source_data)
-            else:
+            elif data in data_args:
                 # Arguments with empty domain do not need to be connected on the nested SDFG.
-                unused_data.add(nsdfg_dataname)
-
-        for data in sorted(unused_data):  # NOTE: remove the data in deterministic order
-            inner_ctx.sdfg.remove_data(data, validate=gtx_config.DEBUG)
+                if (arg_node := data_args[data]) is None:
+                    unused_data.add(data)
+                else:
+                    input_memlets[data] = outer_ctx.sdfg.make_array_memlet(arg_node.dc_node.data)
+            else:
+                # Capture GTIR symbols (scalars, arrays) from the parent scope.
+                assert data in outer_ctx.sdfg.arrays
+                # Ensure that connectivity tables are non-transient arrays in parent SDFG.
+                if data in connectivity_arrays:
+                    outer_ctx.sdfg.arrays[data].transient = False
+                input_memlets[data] = outer_ctx.sdfg.make_array_memlet(data)
+
+        if unused_data:
+            for data in sorted(unused_data):  # NOTE: remove the data in deterministic order
+                assert not inner_ctx.sdfg.arrays[data].transient
+                input_nodes_to_remove = [
+                    access_node
+                    for access_node in inner_ctx.state.data_nodes()
+                    if access_node.data == data
+                ]
+                non_isolated_nodes = [
+                    input_node
+                    for input_node in input_nodes_to_remove
+                    if inner_ctx.state.in_degree(input_node) != 0
```

Comment:
```suggestion
                    if inner_ctx.state.degree(input_node) != 0
```
This will also consider the case were something reads from a node that is supposed to be unused.